torchtune/datasets/_slimorca.py

```diff
@@ -4,53 +4,54 @@
 # This source code is licensed under the BSD-style license found in the
 # LICENSE file in the root directory of this source tree.
 
-from typing import Optional
+from typing import Dict, Optional, Union
 
-from torchtune.datasets._chat import chat_dataset, ChatDataset
+from torchtune.data import PromptTemplate, ShareGPTToMessages
 
-from torchtune.modules.tokenizers import ModelTokenizer
+from torchtune.datasets._finetune import SFTDataset
+from torchtune.datasets._packed import PackedDataset
+from torchtune.modules.transforms import Transform
 
 
 def slimorca_dataset(
-    tokenizer: ModelTokenizer,
+    model_transform: Transform,
     *,
     source: str = "Open-Orca/SlimOrca-Dedup",
-    chat_format: Optional[str] = None,
-    max_seq_len: int = 1024,
+    column_map: Optional[Dict[str, str]] = None,
+    prompt_template: Optional[PromptTemplate] = None,
     train_on_input: bool = False,
     packed: bool = False,
     split: str = "train",
-) -> ChatDataset:
+) -> Union[SFTDataset, PackedDataset]:
     """
     Support for `SlimOrca-style <https://huggingface.co/datasets/Open-Orca/SlimOrca-Dedup>`_
     family of conversational datasets.
 
-    Use a chat format if the base model requires it, such as Llama2 and Mistral.
-    The Llama3 models do not prescribe a particular format.
-
-    The returned data is a tuple of input token id list and label token id
-    list. If ``max_seq_len`` keyword argument is provided, the returned
-    input token id list is ensured (by truncation if necessary) to be within
-    that length.
+    Masking of the prompt during training is controlled by the ``train_on_input`` flag, which is
+    set to ``False`` by default
+    - If ``train_on_input`` is True, the prompt is used during training and
+    contributes to the loss.
+    - If ``train_on_input`` is False, the prompt is masked out (tokens replaced with -100)
 
     Args:
-        tokenizer (ModelTokenizer): Tokenizer used by the model that implements the ``tokenize_messages`` method.
-        source (str): path string of dataset, anything supported by Hugging Face's `load_dataset`.
-        chat_format (Optional[str]): name of template used to format the chat. See the description
-            in :class:`~torchtune.datasets.ChatDataset` for more details. Default: None
-        max_seq_len (int): Maximum number of tokens in the returned input and label token id lists.
-            This value needs to be at least 4 though it is generally set to max sequence length accepted by the model.
-            Default is 1024.
+        model_transform (Transform): model specific transform to convert a list of messages
+            output by the dataset to tokens. This will always be a :class:`~torchtune.modules.tokenizers.ModelTokenizer`.
```

Comment:
I commented this on the other PR (after it landed), but can you clarify why this always has to be a `ModelTokenizer`? And if that's the case, why don't we type it as such?

Reply:
Discussed offline, keeping this as Transform to maintain consistency across SFTDataset, text dataset builders, multimodal dataset builders